### PR TITLE
tests: make the nested library usable independently of spread

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The idea of this change is be able to:

export BUILD_SNAPD_FROM_CURRENT=false 
export UPDATE_PC_KERNEL=false
export BUILD_NEW_IMAGE=true

. tests/lib/nested.sh

create_nested_core_vm
start_nested_core_vm
destroy_nested_vm

Then this lib can be use from other projects and manually to create a vms.

Also now the nested library is prepared to create more that 1 vms to run tests in parallel

### CODE REVIEW

**Comment by stolowski** [comment]
> The idea of this change is be able to:
> 
> export BUILD_SNAPD_FROM_CURRENT=false
> ...

Could you please add this instruction to HACKING.md, under Testing section (I think even in a separate section, or subsection)? I'm not sure if there is a better place than HACKING.md but if there is, feel free to put it there.

I think this needs to be documented somewhere, it's not apparent how to use it otherwise.

As an extra, tests/lib/nested.sh could print a short usage blurb if run/sourced outside of spread ($SPREAD* variable not set).

**Comment by stolowski** [comment]
> > The idea of this change is be able to:
> > export BUILD_SNAPD_FROM_CURRENT=false
> > ...
> 
> Could you please add this instruction to HACKING.md, under Testing section (I think even in a separate section, or subsection)? I'm not sure if there is a better place than HACKING.md but if there is, feel free to put it there.
> 
> I think this needs to be documented somewhere, it's not apparent how to use it otherwise.
> 
> As an extra, tests/lib/nested.sh could print a short usage blurb if run/sourced outside of spread ($SPREAD* variable not set).

Or even better, print usage blurb only if SPREAD or none of the critical variables (such as UPDATE_PC_KERNEL) are set.

**Comment by sergiocazzolato** [comment]
@stolowski sure, I'll add the documentatino, thanks for the suggestion!

File: tests/lib/nested.sh [hunk 1]
```diff
@@ -405,17 +457,26 @@ start_nested_core_vm(){
...
-        # Check the image
-        local IMAGE
+        if [ -z "$CUSTOM_IMAGE_URL" ]; then
+            # create the ubuntu cloud image
+            local IMAGE_URL
+            IMAGE_URL=$(get_image_url_for_nested_vm)
+            wget -P "$WORK_DIR/image" "$IMAGE_URL"
+            # Check the image
+            local IMAGE
         IMAGE=$(ls "$WORK_DIR"/image/*.img)
```
Comment (stolowski) [inline tests/lib/nested.sh:474]: Is indentation wrong in the IMAGE=... line, or is github confused?

File: tests/lib/nested.sh [hunk 2]
```diff
@@ -1,13 +1,69 @@
...
+    echo ""
+    echo "5. Connect to the VM"
+    echo "$ execute_remote"
+    echo ""
+    echo "6. Destroy the VM and clean env"
+    echo "$ destroy_nested_vm"
+    echo "$ cleanup_nested_env"
+}
+
+show_help
```
Comment (stolowski) [inline tests/lib/nested.sh:62]: This should be conditional, no? As is, every test that includes nested.sh will print help when running?

File: tests/lib/nested.sh [hunk 3]
```diff
@@ -1,13 +1,69 @@
...
+NESTED_ARCHITECTURE=${NESTED_ARCHITECTURE:-"amd64"}
+NESTED_TYPE=${NESTED_TYPE:-"core"}
+NESTED_SMP=${NESTED_SMP:-"1"}
+NESTED_MEM=${NESTED_MEM:-"4096"}
+
+# Define defaults for environment variables
+SPREAD_SYSTEM=${SPREAD_SYSTEM:-"$(lsb_release -is | tr '[:upper:]' '[:lower:]')-$(lsb_release -rs)-64"}
+SPREAD_BACKEND=${SPREAD_BACKEND:-"google"}
+
+show_help() {
```
Comment (stolowski) [inline tests/lib/nested.sh:35]: Thanks for adding this, that's going to be useful!

File: tests/lib/nested.sh [hunk 4]
```diff
@@ -1,13 +1,69 @@
...
+    echo "Nested library usage:"
+    echo ""
+    echo "1. Define the variables used to create the VM:"
```
Comment (stolowski) [inline tests/lib/nested.sh:38]: I think they are optional, is that right? Can you expand this help detailing what are the defaults if not provided?

File: tests/lib/nested.sh [hunk 5]
```diff
@@ -1,13 +1,69 @@
...
+    echo "$ export BUILD_SNAPD_FROM_CURRENT=false"
+    echo "$ export UPDATE_PC_KERNEL=false"
+    echo "$ export BUILD_NEW_IMAGE=true"
+    echo "$ export EMULATE_TPM"
```
Comment (stolowski) [inline tests/lib/nested.sh:42]: This is missing a value?

File: tests/lib/nested.sh [hunk 6]
```diff
@@ -1,13 +1,69 @@
...
+    echo ""
+    echo "3. Create the VM and start it as a systemd unit"
+    echo "$ create_nested_core_vm"
+    echo "$ start_nested_core_vm"
+    echo ""
+    echo "4. Check VM status"
+    echo "$ systemctl status nested-vm"
+    echo ""
+    echo "5. Connect to the VM"
+    echo "$ execute_remote"
```
Comment (stolowski) [inline tests/lib/nested.sh:55]: This needs args, or are they optional? Should it say `execute_remote [command]` ?

**Comment by sergiocazzolato** [comment]
This will be split in more than 1 change

**Pull request closed without merging**